File: 07-06-2025_Lexicographically_Minimum_String_After_Removing_Stars.cpp
class Solution {
public:
    string clearStars(string s) {
        

        int n = s.size();
        priority_queue<char, vector<char>, greater<char>> pq;

        unordered_map<char,vector<int>> indexmap;

        vector<bool> v (n,true);
     
        for(int i=0;i<n;i++)
        {
             if(s[i]!='*')
             {
                 pq.push(s[i]);
                 indexmap[s[i]].push_back(i);
              
            }

            else
            {
                 char currentmin = pq.top();
                 pq.pop();

                 int lastpos = indexmap[currentmin].back();
            
                 indexmap[currentmin].pop_back();

                 //dont include the star
                 v[i] = false;
                 //dont include char at the occurence we had to remove

                 v[lastpos] = false;
            }
        }


        string result = "";

        for(int i=0;i<n;i++)
        {
             if(v[i])
             result += s[i];
        }


        return result;

        
    }
};

File: 07-06-2025_Lexicographically_Minimum_String_After_Removing_Stars.java
class Solution {
    public String clearStars(String s) {
        int n = s.length();
        PriorityQueue<Character> pq = new PriorityQueue<>();
        Map<Character, Deque<Integer>> indexMap = new HashMap<>();
        boolean[] keep = new boolean[n];
        Arrays.fill(keep, true);

        for (int i = 0; i < n; i++) {
            char ch = s.charAt(i);
            if (ch != '*') {
                pq.add(ch);
                indexMap.putIfAbsent(ch, new ArrayDeque<>());
                indexMap.get(ch).addLast(i);
            } else {
                char currentMin = pq.poll();
                int lastPos = indexMap.get(currentMin).removeLast();

                keep[i] = false;
                keep[lastPos] = false;

                if (indexMap.get(currentMin).isEmpty()) {
                    indexMap.remove(currentMin);  // Optional cleanup
                }
            }
        }

        StringBuilder result = new StringBuilder();
        for (int i = 0; i < n; i++) {
            if (keep[i]) {
                result.append(s.charAt(i));
            }
        }

        return result.toString();
    }
}
